File: backend/bastion/api/serializers/briefings.py
"""
Briefing and Notification serializers
"""

from rest_framework import serializers
from bastion.briefings.models import BriefingTemplate, Briefing, Notification


class BriefingTemplateSerializer(serializers.ModelSerializer):
    """Serializer for BriefingTemplate model"""
    template_type_display = serializers.CharField(
        source='get_template_type_display', read_only=True
    )
    usage_count = serializers.SerializerMethodField()

    class Meta:
        model = BriefingTemplate
        fields = [
            'id', 'name', 'template_type', 'template_type_display',
            'description', 'subject_template', 'body_template',
            'available_variables', 'is_active', 'requires_approval',
            'schedule_cron', 'usage_count',
            'created_at', 'updated_at'
        ]
        read_only_fields = ['id', 'created_at', 'updated_at']

    def get_usage_count(self, obj):
        return obj.briefings.count()


class BriefingTemplateListSerializer(serializers.ModelSerializer):
    """Lightweight serializer for listing templates"""
    template_type_display = serializers.CharField(
        source='get_template_type_display', read_only=True
    )

    class Meta:
        model = BriefingTemplate
        fields = [
            'id', 'name', 'template_type', 'template_type_display',
            'is_active', 'requires_approval'
        ]


class BriefingSerializer(serializers.ModelSerializer):
    """Full serializer for Briefing model"""
    household_name = serializers.CharField(source='household.name', read_only=True)
    client_name = serializers.CharField(source='client.full_name', read_only=True)
    template_name = serializers.CharField(source='template.name', read_only=True)
    created_by_name = serializers.CharField(source='created_by.full_name', read_only=True)
    approved_by_name = serializers.CharField(source='approved_by.full_name', read_only=True)
    status_display = serializers.CharField(source='get_status_display', read_only=True)
    delivery_method_display = serializers.CharField(
        source='get_delivery_method_display', read_only=True
    )
    attachment_count = serializers.SerializerMethodField()

    class Meta:
        model = Briefing
        fields = [
            'id', 'title', 'subject',
            'household', 'household_name', 'client', 'client_name',
            'template', 'template_name',
            'body_markdown', 'body_html',
            'status', 'status_display',
            'delivery_method', 'delivery_method_display',
            'scheduled_for', 'sent_at', 'opened_at',
            'created_by', 'created_by_name',
            'approved_by', 'approved_by_name', 'approved_at',
            'period_start', 'period_end',
            'attachment_count',
            'created_at', 'updated_at'
        ]
        read_only_fields = [
            'id', 'body_html', 'sent_at', 'opened_at',
            'created_by', 'approved_at', 'created_at', 'updated_at'
        ]

    def get_attachment_count(self, obj):
        return obj.attachments.count()


class BriefingListSerializer(serializers.ModelSerializer):
    """Lightweight serializer for listing briefings"""
    recipient = serializers.SerializerMethodField()
    status_display = serializers.CharField(source='get_status_display', read_only=True)

    class Meta:
        model = Briefing
        fields = [
            'id', 'title', 'recipient', 'status', 'status_display',
            'delivery_method', 'scheduled_for', 'sent_at', 'created_at'
        ]

    def get_recipient(self, obj):
        if obj.client:
            return obj.client.full_name
        if obj.household:
            return obj.household.name
        return 'Unknown'


class BriefingCreateSerializer(serializers.ModelSerializer):
    """Serializer for creating briefings"""

    class Meta:
        model = Briefing
        fields = [
            'title', 'subject', 'household', 'client',
            'template', 'body_markdown',
            'delivery_method', 'scheduled_for',
            'period_start', 'period_end'
        ]

    def validate(self, attrs):
        if not attrs.get('household') and not attrs.get('client'):
            raise serializers.ValidationError(
                'Either household or client must be specified.'
            )
        return attrs

    def create(self, validated_data):
        request = self.context.get('request')
        if request and request.user.is_authenticated:
            validated_data['created_by'] = request.user
        return super().create(validated_data)


class NotificationSerializer(serializers.ModelSerializer):
    """Serializer for Notification model"""
    notification_type_display = serializers.CharField(
        source='get_notification_type_display', read_only=True
    )

    class Meta:
        model = Notification
        fields = [
            'id', 'title', 'message',
            'notification_type', 'notification_type_display',
            'link', 'link_text',
            'is_read', 'read_at',
            'created_at'
        ]
        read_only_fields = ['id', 'created_at']


class NotificationListSerializer(serializers.ModelSerializer):
    """Lightweight serializer for listing notifications"""

    class Meta:
        model = Notification
        fields = [
            'id', 'title', 'notification_type',
            'is_read', 'created_at'
        ]


class NotificationMarkReadSerializer(serializers.Serializer):
    """Serializer for marking notifications as read"""
    notification_ids = serializers.ListField(
        child=serializers.UUIDField(),
        required=False
    )
    mark_all = serializers.BooleanField(default=False)

File: frontend/src/lib/api/briefings.ts
import { api } from './client'
import type { PaginatedResponse } from './clients'

export interface BriefingTemplate {
  id: string
  name: string
  template_type: string
  template_type_display: string
  description: string
  subject_template: string
  body_template: string
  available_variables: string[]
  is_active: boolean
  requires_approval: boolean
  usage_count: number
  created_at: string
}

export interface Briefing {
  id: string
  title: string
  subject: string
  household: string | null
  household_name: string | null
  client: string | null
  client_name: string | null
  template: string | null
  template_name: string | null
  body_markdown: string
  body_html: string
  status: 'draft' | 'pending_review' | 'approved' | 'sent' | 'failed'
  status_display: string
  delivery_method: 'email' | 'portal' | 'both'
  delivery_method_display: string
  scheduled_for: string | null
  sent_at: string | null
  opened_at: string | null
  created_by: string | null
  created_by_name: string | null
  approved_by: string | null
  approved_by_name: string | null
  approved_at: string | null
  period_start: string | null
  period_end: string | null
  attachment_count: number
  created_at: string
  updated_at: string
}

export interface Notification {
  id: string
  title: string
  message: string
  notification_type: 'info' | 'warning' | 'alert' | 'success' | 'task'
  notification_type_display: string
  link: string
  link_text: string
  is_read: boolean
  read_at: string | null
  created_at: string
}

export const briefingsApi = {
  async list(params?: Record<string, string>): Promise<PaginatedResponse<Briefing>> {
    const queryString = params ? '?' + new URLSearchParams(params).toString() : ''
    return api.get(`/briefings/${queryString}`)
  },

  async get(id: string): Promise<Briefing> {
    return api.get(`/briefings/${id}/`)
  },

  async create(data: Partial<Briefing>): Promise<Briefing> {
    return api.post('/briefings/', data)
  },

  async update(id: string, data: Partial<Briefing>): Promise<Briefing> {
    return api.patch(`/briefings/${id}/`, data)
  },

  async approve(id: string): Promise<Briefing> {
    return api.post(`/briefings/${id}/approve/`)
  },

  async send(id: string): Promise<Briefing> {
    return api.post(`/briefings/${id}/send/`)
  },

  async getPending(): Promise<Briefing[]> {
    return api.get('/briefings/pending/')
  },

  async getScheduled(): Promise<Briefing[]> {
    return api.get('/briefings/scheduled/')
  },
}

export const briefingTemplatesApi = {
  async list(): Promise<PaginatedResponse<BriefingTemplate>> {
    return api.get('/briefing-templates/')
  },

  async get(id: string): Promise<BriefingTemplate> {
    return api.get(`/briefing-templates/${id}/`)
  },
}

export const notificationsApi = {
  async list(params?: Record<string, string>): Promise<PaginatedResponse<Notification>> {
    const queryString = params ? '?' + new URLSearchParams(params).toString() : ''
    return api.get(`/notifications/${queryString}`)
  },

  async getUnread(): Promise<{ count: number; notifications: Notification[] }> {
    return api.get('/notifications/unread/')
  },

  async markRead(notificationIds?: string[]): Promise<{ marked_read: number }> {
    if (notificationIds) {
      return api.post('/notifications/mark_read/', { notification_ids: notificationIds })
    }
    return api.post('/notifications/mark_read/', { mark_all: true })
  },

  async markOneRead(id: string): Promise<Notification> {
    return api.post(`/notifications/${id}/read/`)
  },
}
